src/livekit/useLiveKit.ts

```diff
@@ -53,25 +60,38 @@ interface UseLivekitResult {
 
 export function useLiveKit(
   muteStates: MuteStates,
+  rtcSession: MatrixRTCSession,
   sfuConfig?: SFUConfig,
   e2eeConfig?: E2EEConfig
 ): UseLivekitResult {
   const e2eeOptions = useMemo(() => {
-    if (!e2eeConfig?.sharedKey) return undefined;
+    if (!e2eeConfig) return undefined;
 
-    return {
-      keyProvider: new ExternalE2EEKeyProvider(),
-      worker: new E2EEWorker(),
-    } as E2EEOptions;
+    if (e2eeConfig.mode === E2EEMode.PerParticipantKey) {
+      return {
+        keyProvider: new MatrixKeyProvider(),
+        worker: new E2EEWorker(),
+      };
+    } else if (e2eeConfig.mode === E2EEMode.SharedKey && e2eeConfig.sharedKey) {
+      return {
+        keyProvider: new ExternalE2EEKeyProvider(),
+        worker: new E2EEWorker(),
+      };
+    }
   }, [e2eeConfig]);
 
   useEffect(() => {
-    if (!e2eeConfig?.sharedKey || !e2eeOptions) return;
-
-    (e2eeOptions.keyProvider as ExternalE2EEKeyProvider).setKey(
-      e2eeConfig?.sharedKey
-    );
-  }, [e2eeOptions, e2eeConfig?.sharedKey]);
+    if (!e2eeOptions) return;
+    if (!e2eeConfig) return;
+
+    if (e2eeConfig.mode === E2EEMode.PerParticipantKey) {
+      (e2eeOptions.keyProvider as MatrixKeyProvider).setRTCSession(rtcSession);
```

Comment:
Should `setRTCSession` be called with `undefined` somewhere to remove the listener?

Reply:
I assume that if the class gets destroyed, so do the listeners, so there is no need to call it with `undefined`?

Reply:
Well, not explicitly: there's no destructors run here (it's javascript) so the emitter will continue to retain a reference to the listener, we'd just be relying on them being GCed, so it depends whether anything else holds on to the thing doing the listening (the key provider in this case I think). Assuming everything gets torn down then everything should be detached from the GC root, but we shouldn't be relying on it given the choice IMO.